lighthouse-core/scripts/i18n/collection-util.js

```diff
@@ -0,0 +1,281 @@
+
+/**
+ * @typedef ICUPlaceholderDefn
+ * @property {string} content
+ * @property {string} [example]
+ */
+
+// @ts-ignore - @types/esprima lacks all of these
+function computeDescription(ast, property, value, startRange) {
+  const endRange = property.range[0];
+  for (const comment of ast.comments || []) {
+    if (comment.range[0] < startRange) continue;
+    if (comment.range[0] > endRange) continue;
+    if (comment.value.includes('@')) {
+      // This is a complex description with description and examples.
+      let description = '';
+      /** @type {Record<string, string>} */
+      const examples = {};
+
+      const r = /@(\w+) ({\w+})?(.*)(\n|$)/g;
+      let resArr;
+      while ((resArr = r.exec(comment.value)) !== null) {
+        const tagName = resArr[1];
+        const placeholder = resArr[2];
+        const message = resArr[3].trim();
+
+        if (tagName === 'description') {
+          description = message;
+        } else if (tagName === 'example') {
+          // Make sure the ICU var exists in string
+          if (!value.includes(placeholder)) {
+            throw Error('Example missing ICU replacement');
+          }
+          examples[placeholder.substring(1, placeholder.length - 1)] = message;
+        }
+      }
+      // Make sure all ICU vars have examples
+      const rMini = /\{(\w+)\}/g;
+      while ((resArr = rMini.exec(value)) !== null) {
+        const varName = resArr[1];
+        if (!examples[varName]) throw Error(`Variable '${varName}' is missing example comment`);
+      }
+      return {description, examples};
+    }
+    return {description: comment.value.replace('*', '').trim()};
+  }
+  return {};
+}
+
+/**
+ * @param {*} message
+ * @param {*} examples
+ * @returns {ICUMessageDefn}
+ */
+function convertMessageToPlaceholders(message, examples = {}) {
+  // console.log(examples);
+  // Basically the same as markdown parsing in dom.js
+  /** @type {Record<string, ICUPlaceholderDefn>} */
+  const placeholders = {};
+
+  // Sanity checks
+  // Number of backticks is even.
+  if ((message.split('`').length - 1) % 2 !== 0) {
+    throw Error(`Open backtick in message "${message}"`);
+  }
+  // Link markdown is not slightly off.
+  if (message.match(/\[.*\] \(.*\)/)) {
+    throw Error(`Bad Link syntax in message "${message}"`);
+  }
+  // Complex ICU using non-supported format
+  if (message.match(
+    /\{(\w{2,50}), number, (?!milliseconds|seconds|bytes|percent|extendedPercent).*\}/)) {
+    throw Error(`Non supported ICU format in message "${message}"`);
+  }
+
+  // replace code snippets
+  let parts = message.split(/`(.*?)`/g); // Split on markdown code slashes
+  let newMessage = '';
+  let idx = 0;
+  while (parts.length) {
+    // Pop off the same number of elements as there are capture groups.
+    const [preambleText, codeText] = parts.splice(0, 2);
+    newMessage += preambleText;
+    if (codeText) {
+      const pName = `MARKDOWN_SNIPPET_${idx++}`;
+      newMessage += `$${pName}$`;
```

Comment:
this is a fun snippet to read and it only gets better below :)

`['$', pname, '$'].join('')` almost feels better 😆

Reply:
Haha, I will admit, this isn't pretty.